In [1]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install pymongo

Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas
import pymongo

URI = "mongodb+srv://user_mongo:ymHlgu61PHcPpecr@cluster-but-sd.hz0y4.mongodb.net/?retryWrites=true&w=majority&appName=Cluster-but-sd"
client = pymongo.MongoClient(URI)
db = client.sae

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())
print("SAE: ", db.sae)

C:\Users\j.fond\AppData\Local\anaconda3\Lib\site-packages\pymongo\pyopenssl_context.py:345: CryptographyDeprecationWarning: Parsed a negative serial number, which is disallowed by RFC 5280. Loading this certificate will cause an exception in the next release of cryptography.
  _crypto.X509.from_cryptography(x509.load_der_x509_certificate(cert))


Collections:  ['orders', 'employees', 'customers', 'products']
SAE:  Collection(Database(MongoClient(host=['cluster-but-sd-shard-00-02.hz0y4.mongodb.net:27017', 'cluster-but-sd-shard-00-00.hz0y4.mongodb.net:27017', 'cluster-but-sd-shard-00-01.hz0y4.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', appname='Cluster-but-sd', authsource='admin', replicaset='atlas-11vlh4-shard-0', tls=True), 'sae'), 'sae')


In [2]:
d = db.orders.find(limit = 1)
d
list(d)

[{'_id': ObjectId('675c0406fdc9451b21860b8f'),
  'orderNumber': 10103,
  'orderDate': '2003/1/29 0:00:00',
  'requiredDate': '2003/2/7 0:00:00',
  'shippedDate': '2003/2/2 0:00:00',
  'status': 'Shipped',
  'comments': 'NULL',
  'customerNumber': 121,
  'order_details': [{'productCode': 'S10_1949',
    'quantityOrdered': 26,
    'priceEach': 207.87,
    'orderLineNumber': 11},
   {'productCode': 'S10_4962',
    'quantityOrdered': 42,
    'priceEach': 128.53,
    'orderLineNumber': 4},
   {'productCode': 'S12_1666',
    'quantityOrdered': 27,
    'priceEach': 125.74,
    'orderLineNumber': 8},
   {'productCode': 'S18_1097',
    'quantityOrdered': 35,
    'priceEach': 112.0,
    'orderLineNumber': 10},
   {'productCode': 'S18_2432',
    'quantityOrdered': 22,
    'priceEach': 54.09,
    'orderLineNumber': 2},
   {'productCode': 'S18_2949',
    'quantityOrdered': 27,
    'priceEach': 83.07,
    'orderLineNumber': 12},
   {'productCode': 'S18_2957',
    'quantityOrdered': 35,
    'priceEac

### 1. Lister les clients n’ayant jamais effecuté une commande

In [134]:
pandas.DataFrame(list(db.customers.aggregate([
    { "$lookup": { #Jointure entre la table "orders" et "customers"
       "from": "orders",
       "localField": "customerNumber",
       "foreignField": "customerNumber",
       "as": "Orders"}},
    {"$match" : {"Orders" : {"$size" : 0}}}
])))

_id  customerNumber                    customerName  \
0   675c047cfdc9451b21860cef             125              Havel & Zbyszek Co   
1   675c047cfdc9451b21860cfd             168          American Souvenirs Inc   
2   675c047cfdc9451b21860cfe             169               Porto Imports Co.   
3   675c047cfdc9451b21860d19             247        Messner Shopping Network   
4   675c047cfdc9451b21860d28             307                Der Hund Imports   
5   675c047cfdc9451b21860d37             348           Asian Treasures, Inc.   
6   675c047cfdc9451b21860d3a             356            SAR Distributors, Co   
7   675c047cfdc9451b21860d5a             477             Mit Vergnügen & Co.   
8   675c047cfdc9451b21860d5c             481              Raanan Stores, Inc   
9   675c047cfdc9451b21860d15             237                   ANG Resellers   
10  675c047cfdc9451b21860d24             293               BG&E Collectables   
11  675c047cfdc9451b21860d3c             361                 Kommission Auto   
12  675c047cfdc9451b21860d4c             443        Feuer Online Stores, Inc   
13  675c047cfdc9451b21860d54             459                Warburg Exchange   
14  675c047cfdc9451b21860d5b             480       Kremlin Collectables, Co.   
15  675c047cfdc9451b21860d0d             206      Asian Shopping Network, Co   
16  675c047cfdc9451b21860d27             303                Schuyler Imports   
17  675c047cfdc9451b21860d48             409  Stuttgart Collectable Exchange   
18  675c047cfdc9451b21860d56             465             Anton Designs, Ltd.   
19  675c047cfdc9451b21860d12             223                 Natürlich Autos   
20  675c047cfdc9451b21860d1f             273               Franken Gifts, Co   
21  675c047cfdc9451b21860d33             335       Cramer Spezialitäten, Ltd   
22  675c047cfdc9451b21860d3f             369          Lisboa Souveniers, Inc   
23  675c047cfdc9451b21860d40             376           Precious Collectables   

    contactLastName contactFirstName             phone  \
0   Piestrzeniewicz          Zbyszek     (26) 642-7555   
1            Franco              Sue        2035557845   
2         de Castro           Isabel      (1) 356-5555   
3           Messner           Renate       069-0555984   
4            Anders            Maria       030-0074555   
5           McKenna         Patricia          2967 555   
6             Kuger           Armand   +27 21 550 3555   
7              Moos            Hanna        0621-08555   
8       Altagar,G M           Raanan  + 972 9 959 8555   
9            Camino        Alejandra     (91) 745 6555   
10              Pon               Ed  +41 26 425 50 01   
11          Josephs            Karin       0251-555259   
12            Feuer        Alexander       0342-555176   
13          Ottlieb             Sven       0241-039123   
14          Semenov        Alexander   +7 812 293 0521   
15           Walker           Brydey    +612 9411 1555   
16         Schuyler          Bradley   +31 20 491 9555   
17           Müller             Rita       0711-555361   
18            Anton           Carmen    +34 913 728555   
19            Kloss            Horst       0372-555188   
20          Franken            Peter       089-0877555   
21           Cramer           Philip        0555-09555   
22        Rodriguez             Lino      (1) 354-2555   
23             Wang             Yang       0452-076555   

                                      addressLine1 addressLine2  \
0                                  ul. Filtrowa 68         NULL   
1                                149 Spinnaker Dr.    Suite 101   
2                           Estrada da saúde n. 58         NULL   
3                                     Magazinweg 7         NULL   
4                                    Obere Str. 57         NULL   
5                                 8 Johnstown Road         NULL   
6       Century City-Montague Gardens PO Box 37177         NULL   
7                                   F

### 2. Pour chaque employé, le nombre de clients, le nombre de commandes et le montant total de celles-ci

In [90]:
pandas.DataFrame(list(db.employees.aggregate([
    {"$lookup": { #Jointure entre la table "employees" et "customers" 
        "from": "customers",
        "localField": "employeeNumber",
        "foreignField": "salesRepEmployeeNumber",
        "as": "clients"}},
    {"$unwind": {"path": "$clients", "preserveNullAndEmptyArrays": True}}, #séparation des documents en gardant les NA
    
    {"$lookup": { #Jointure entre la table "orders" et "customers" 
        "from": "orders",
        "localField": "clients.customerNumber",
        "foreignField": "customerNumber",
        "as": "orders"}},
    
    {"$unwind": {"path": "$orders", "preserveNullAndEmptyArrays": True}}, 
    
    {"$unwind": {"path": "$orders.payments", "preserveNullAndEmptyArrays": True}},
    
    {"$group": { #regroupement par numéro d'employé
        "_id": "$employeeNumber",
        "firstname": {"$first": "$firstName"},  
        "lastname": {"$first": "$lastName"}, 
        "nombre_clients": {"$addToSet": "$clients.customerNumber"}, 
        "commandes_uniques": {"$addToSet": "$orders.orderNumber"},
        "montant_total": {"$sum": {"$ifNull": ["$orders.payments.amount", 0]}}}},
    
    {"$project": {
        "firstname": 1,
        "lastname": 1,
        "nombre_clients": {"$size": "$nombre_clients"}, 
        "nombre_commandes": {"$size": "$commandes_uniques"},
        "montant_total": 1}}
])))

_id firstname   lastname  montant_total  nombre_clients  nombre_commandes
0   1002     Diane     Murphy           0.00               0                 0
1   1102    Gerard     Bondur           0.00               0                 0
2   1501     Larry       Bott     2188718.80               8                22
3   1370    Gerard  Hernandez    21850743.39               7                43
4   1056      Mary  Patterson           0.00               0                 0
5   1286  Foon Yue      Tseng     1635226.60               7                17
6   1216     Steve  Patterson     1545990.08               6                18
7   1188     Julie   Firrelli     1035043.99               6                14
8   1337      Loui     Bondur     2093279.51               6                20
9   1166    Leslie   Thompson      943442.48               6                14
10  1076      Jeff   Firrelli           0.00               0                 0
11  1619       Tom       King           0.00               0                 0
12  1401    Pamela   Castillo     2580550.55              10                31
13  1504     Barry      Jones     2357369.87               9                25
14  1612     Peter      Marsh     2119701.72               5                19
15  1143   Anthony        Bow           0.00               0                 0
16  1165    Leslie   Jennings    12674066.13               6                34
17  1323    George     Vanauf     1870933.55               8                22
18  1625   Yoshimi       Kato           0.00               0                 0
19  1611      Andy     Fixter     2352253.84               5                19
20  1621      Mami      Nishi     1815463.17               5                16
21  1702    Martin     Gerard     1014439.14               6                12
22  1088   William  Patterson           0.00               0                 0

### 3. Idem pour chaque bureau (nombre de clients, nombre de commandes et montant total), avec en plus le nombre de clients d’un pays différent, s’il y en a ;

In [128]:
pandas.DataFrame(list(db.employees.aggregate([
    {"$lookup": { #Jointure entre la table "employees" et "customers" 
        "from": "customers",
        "localField": "employeeNumber",
        "foreignField": "salesRepEmployeeNumber",
        "as": "clients"}},
    
    {"$unwind": {"path": "$clients", "preserveNullAndEmptyArrays": True}}, 
    
    {"$lookup": {
        "from": "orders",
        "localField": "clients.customerNumber",
        "foreignField": "customerNumber",
        "as": "orders"}},
    
    {"$unwind": {"path": "$orders", "preserveNullAndEmptyArrays": True}},

    {"$unwind": {"path": "$orders.payments", "preserveNullAndEmptyArrays": True}},

    {"$addFields": {
        "client_etranger": {
            "$cond": {
                "if": {"$ne": ["$clients.country", "$office.country"]},
                "then": 1,
                "else": 0}}}},

    {"$group": { #regroupement par code des bureaux
        "_id": "$officeCode",
        "officeCity": {"$first": "$office.city"},  
        "officeCountry": {"$first": "$office.country"}, 
        "nombre_clients": {"$addToSet": "$clients.customerNumber"},
        "nombre_commandes": {"$addToSet": "$orders.orderNumber"},
        "montant_total": {"$sum": {"$ifNull": ["$orders.payments.amount", 0]}},
        "nombre_clients_etrangers": {"$sum": "$client_etranger"}}},

    {"$project": {
        "_id": 0,
        "_id": 1,
        "officeCity" : 1,
        "officeCountry" : 1,
        "nombre_clients": {"$size": "$nombre_clients"},
        "nombre_commandes": {"$size": "$nombre_commandes"},
        "montant_total": 1,
        "nombre_clients_etrangers": 1}}
])))

_id       officeCity officeCountry  montant_total  \
0  1.0  [San Francisco]         [USA]    13617508.61   
1  5.0          [Tokyo]       [Japan]     1815463.17   
2  4.0          [Paris]      [France]    27539012.59   
3  6.0         [Sydney]   [Australia]     4471955.56   
4  2.0         [Boston]         [USA]     2581034.07   
5  3.0            [NYC]         [USA]     3506160.15   
6  7.0         [London]          [UK]     4546088.67   

   nombre_clients_etrangers  nombre_clients  nombre_commandes  
0                       247              12                48  
1                        48               5                16  
2                       566              29               106  
3                       125              10                38  
4                        79              12                32  
5                       107              15                39  
6                       127              17                47

### 4. Pour chaque produit, donner le nombre de commandes, la quantité totale commandée, et le nombre de clients différents (110 lignes)

In [118]:
pandas.DataFrame(list(db.orders.aggregate([
    {"$unwind": "$order_details"},
    
    {"$group": { #regroupement par code produit
        "_id": "$order_details.productCode", 
        "nombre_commandes": {"$addToSet": "$orderNumber"},  
        "quantite_totale": {"$sum": "$order_details.quantityOrdered"}, 
        "clients_uniques": {"$addToSet": "$customerNumber"}}},

    {"$project": {"_id": 0,
                  "productCode": "$_id",
                  "nombre_commandes": {"$size": "$nombre_commandes"},
                  "quantite_totale": 1,
                  "nombre_clients": {"$size": "$clients_uniques"}}}
])))

quantite_totale productCode  nombre_commandes  nombre_clients
0                855    S18_2870                25              21
1                912    S24_2972                27              21
2                972    S18_1889                27              25
3                932    S10_4962                28              25
4               1013    S18_4721                27              21
..               ...         ...               ...             ...
104             1009    S24_1578                28              25
105              933    S12_1099                27              23
106             1028    S12_2823                28              25
107              985    S10_4698                28              25
108              949    S24_2766                25              18

[109 rows x 4 columns]

### 5. Donner le nombre de commande pour chaque pays, ainsi que le montant total des commandes et le montant total payé : on veut conserver les clients n’ayant jamais commandé dans le résultat final

In [132]:
pandas.DataFrame(list(db.customers.aggregate([
    {"$lookup": { #jointure entre "customers" et "orders"
            "from": "orders",
            "localField": "customerNumber",
            "foreignField": "customerNumber",
            "as": "customerOrders"}},

    {"$unwind": {"path": "$customerOrders",
                 "preserveNullAndEmptyArrays": True}},
    {"$unwind": {"path": "$customerOrders.order_details",
                 "preserveNullAndEmptyArrays": True}},
    {"$unwind": {"path": "$customerOrders.payments",
                 "preserveNullAndEmptyArrays": True}},
    
    {"$group": {"_id": {"country": "$country","customerNumber": "$customerNumber"}, #regroupement par pays
                "total_payé": {"$sum": "$customerOrders.payments.amount"},
                "total_commande": {"$sum": {"$multiply": ["$customerOrders.order_details.quantityOrdered","$customerOrders.order_details.priceEach"]}},
                "uniqueOrders": {"$addToSet": "$customerOrders.orderNumber"}}},
    
    {"$addFields": {"nombre_commandes": {"$size": "$uniqueOrders"}}},
   
    {"$group": {"_id": "$_id.country",
                "nombre_commandes": {"$sum": "$nombre_commandes"},
                "total_payé": {"$sum": "$total_payé"},
                "total_commande": {"$sum": "$total_commande"}}},

    {"$project": {"_id": 0,
                  "Pays": "$_id",
                  "Nombre_de_commandes": "$nombre_commandes",
                  "Total_Commandé": "$total_commande",
                  "Total_Payé": "$total_payé"}}
])))

Pays  Nombre_de_commandes  Total_Commandé    Total_Payé
0         France                   37      3160296.75  3.141444e+07
1        Finland                    9       988745.73  1.009620e+07
2        Denmark                    7       781357.50  7.001114e+06
3         Israel                    0            0.00  0.000000e+00
4          Spain                   36     12665636.19  2.123241e+08
5      Australia                   19      2182269.38  2.482541e+07
6      Singapore                    9      1038454.91  1.150336e+07
7         Canada                    7       448157.12  4.487022e+06
8        Belgium                    7       283705.44  1.931535e+06
9          Japan                    6       496898.36  4.837611e+06
10       Ireland                    2       115512.86  9.241029e+05
11  South Africa                    0            0.00  0.000000e+00
12        Sweden                    7       554287.75  6.541197e+06
13       Austria                    7       606187.59  4.090982e+06
14        Norway                    9       848125.78  8.870124e+06
15         Italy                   10       945208.16  1.324310e+07
16        Russia                    0            0.00  0.000000e+00
17            UK                   13      1204373.23  1.506256e+07
18   New Zealand                   15      1736137.04  1.710337e+07
19     Hong Kong                    2        48784.36  7.805498e+05
20           USA                  112     14228729.12  1.957451e+08
21   Philippines                    3       282047.19  2.444409e+06
22        Poland                    0            0.00  0.000000e+00
23   Switzerland                    2       235427.12  3.649120e+06
24      Portugal                    0            0.00  0.000000e+00
25   Netherlands                    0            0.00  0.000000e+00
26       Germany                    7       576293.44  4.971661e+06

### 6. On veut la table de contigence du nombre de commande entre la ligne de produits et le pays du client (127 lignes)

In [113]:
pandas.DataFrame(list(db.customers.aggregate([
    {"$lookup": {"from": "orders", #jointure entre "customers" et "orders"
                 "localField": "customerNumber",
                 "foreignField": "customerNumber",
                 "as": "orders"}},

    {"$unwind": "$orders"},

    {"$unwind": "$orders.order_details"},

    {"$lookup": { #jointure entre "products" et "orders" 
            "from": "products",
            "localField": "orders.order_details.productCode",
            "foreignField": "productCode",
            "as": "product_info"}},

    {"$unwind": "$product_info"},

    {"$group": {"_id": {"ligne_produit": "$product_info.productLine", #regroupement par ligne de produit
                        "pays": "$country"},
                "commandes_uniques": {"$addToSet": "$orders.orderNumber"}}},

    {"$project": {"_id": 0,
                  "Ligne_de_Produit": "$_id.ligne_produit",
                  "Pays": "$_id.pays",
                  "Nombre_de_Commandes": {"$size": "$commandes_uniques"}}},

    {"$sort": {"Ligne_de_Produit": 1, "Pays": 1}} #tri par ordre alphabétique
])))

Ligne_de_Produit       Pays  Nombre_de_Commandes
0       Classic Cars  Australia                   12
1       Classic Cars    Austria                    5
2       Classic Cars    Belgium                    2
3       Classic Cars     Canada                    6
4       Classic Cars    Denmark                    5
..               ...        ...                  ...
121     Vintage Cars  Singapore                    4
122     Vintage Cars      Spain                   22
123     Vintage Cars     Sweden                    4
124     Vintage Cars         UK                   10
125     Vintage Cars        USA                   67

[126 rows x 3 columns]

### 7. On veut la même table croisant la ligne de produits et le pays du client, mais avec le montant total payé dans chaque cellule 

In [83]:
pandas.DataFrame(list(db.customers.aggregate([

    {"$lookup": {"from": "orders", #jointure entre "customers" et "orders" 
                 "localField": "customerNumber",
                 "foreignField": "customerNumber",
                 "as": "orders"}},

    {"$unwind": {"path": "$orders",
                 "preserveNullAndEmptyArrays": False}},

    {"$lookup": {"from": "products", #jointure entre "products" et "orders"
                 "localField": "orders.order_details.productCode",
                 "foreignField": "productCode",
                 "as": "product_info"}},

    {"$unwind": {"path": "$product_info",
                 "preserveNullAndEmptyArrays": False}},

    {"$unwind": {"path": "$orders.payments",
                 "preserveNullAndEmptyArrays": False}},

    {"$group": {"_id": {"ligne_produit": "$product_info.productLine", #regroupement par ligne de produit
                        "pays": "$country"},
                "montant_total_paye": {"$sum": "$orders.payments.amount"}}},

    {"$project": {"_id": 0,
                  "Ligne_de_Produit": "$_id.ligne_produit",
                  "Pays": "$_id.pays",
                  "Montant_Total_Paye": "$montant_total_paye"}},

    {"$sort": {"Ligne_de_Produit": 1, "Pays": 1}} #tri par ordre alphabétique
])))

Ligne_de_Produit       Pays  Montant_Total_Paye
0       Classic Cars  Australia          7504795.97
1       Classic Cars    Austria          1884419.42
2       Classic Cars    Belgium           166880.87
3       Classic Cars     Canada           774924.01
4       Classic Cars    Denmark          3678313.22
..               ...        ...                 ...
121     Vintage Cars  Singapore          1941227.28
122     Vintage Cars      Spain         39878490.66
123     Vintage Cars     Sweden          1377094.16
124     Vintage Cars         UK          4673162.78
125     Vintage Cars        USA         47090713.86

[126 rows x 3 columns]

### 8. Donner les 10 produits pour lesquels la marge moyenne est la plus importante (cf buyPrice et priceEach) ;

In [84]:
pd.DataFrame(list(db.orders.aggregate([

    {"$unwind": {"path": "$order_details",
                 "preserveNullAndEmptyArrays": False}},

    {"$lookup": {"from": "products", #jointure entre "orders" et "products" 
                 "localField": "order_details.productCode",
                 "foreignField": "productCode",
                 "as": "product_info"}},

    {"$unwind": {"path": "$product_info",
                 "preserveNullAndEmptyArrays": False}},

    {"$addFields": {"marge": {"$subtract": ["$order_details.priceEach", "$product_info.buyPrice"]}}}, #ajout de "marge" dans les documents

    {"$group": {"_id": "$order_details.productCode", #regroupement par code produit
                "nom_produit": {"$first": "$product_info.productName"},
                "marge_moyenne": {"$avg": "$marge"}}},

    {"$sort": {"marge_moyenne": -1}}, #tri par ordre décroissant

    {"$limit": 10} #limiter à 10
])))

_id                           nom_produit  marge_moyenne
0  S10_1949              1952 Alpine Renault 1300      99.006429
1  S10_4698  2003 Harley-Davidson Eagle Drag Bike      95.235000
2  S18_3232           1992 Ferrari 360 Spider red      83.334906
3  S12_2823                      2002 Suzuki XREO      83.201429
4  S18_2795                1928 Mercedes-Benz SSK      82.696786
5  S12_1108                     2001 Ferrari Enzo      81.043704
6  S12_3891                      1969 Ford Falcon      77.335926
7  S18_3685        1948 Porsche Type 356 Roadster      72.636800
8  S18_2870          1999 Indy 500 Monte Carlo SS      71.794400
9  S18_1749              1917 Grand Touring Sedan      70.432800

### 9. Lister les produits (avec le nom et le code du client) qui ont été vendus à perte : Si un produit a été dans cette situation plusieurs fois, il doit apparaître plusieurs fois. Une vente à perte arrive quand le prix de vente est inférieur au prix d’achat

In [115]:
pd.DataFrame(list(db.orders.aggregate([
    {"$unwind": {"path": "$order_details",
                 "preserveNullAndEmptyArrays": False}},

    {"$lookup": {"from": "products", #jointure entre "orders" et "products" 
                 "localField": "order_details.productCode",
                 "foreignField": "productCode",
                 "as": "product_info"}},

    {"$unwind": {"path": "$product_info",
                 "preserveNullAndEmptyArrays": False}},
    
    {"$addFields": {"vendu_a_perte": {"$lt": ["$order_details.priceEach", "$product_info.buyPrice"]}}}, #ajout de "vendu_a_perte" dans les documents

    {"$match": {"vendu_a_perte": True}},

    {"$project": {"_id": 0,
                  "code_client": "$customerNumber",
                  "nom_produit": "$product_info.productName",
                  "code_produit": "$order_details.productCode",
                  "prix_vente": "$order_details.priceEach",
                  "prix_achat": "$product_info.buyPrice"}}
])))

code_client                              nom_produit code_produit  \
0           363                   1962 LanciaA Delta 16V     S10_4962   
1           363                       1934 Ford V8 Coupe     S18_2957   
2           363      18th Century Vintage Horse Carriage     S18_3136   
3           145                   1999 Yamaha Speed Boat     S18_3029   
4           129                   1940 Ford Pickup Truck     S18_1097   
..          ...                                      ...          ...   
74          311  1941 Chevrolet Special Deluxe Cabriolet     S18_3856   
75          462                    1996 Moto Guzzi 1100i     S10_2016   
76          462     2003 Harley-Davidson Eagle Drag Bike     S10_4698   
77          448                       1968 Dodge Charger     S12_3380   
78          323           1998 Chrysler Plymouth Prowler     S18_2238   

    prix_vente  prix_achat  
0        61.99      103.42  
1        29.87       34.35  
2        47.04       60.74  
3        45.86       51.61  
4        40.25       58.33  
..         ...         ...  
74       60.49       64.58  
75       44.51       68.99  
76       86.77       91.02  
77       72.38       75.16  
78       69.81      101.51  

[79 rows x 5 columns]

### 10. (bonus) Lister les clients pour lesquels le montant total payé est supérieur aux montants totals des achats

In [133]:
pandas.DataFrame(list(db.customers.aggregate([

    {"$lookup": {"from": "orders", #jointure entre "customers" et "orders"
                 "localField": "customerNumber",
                 "foreignField": "customerNumber",
                 "as": "orders"}},

    {"$unwind": {"path": "$orders",
                 "preserveNullAndEmptyArrays": True}},

    {"$unwind": {"path": "$orders.order_details",
                 "preserveNullAndEmptyArrays": True}},

    {"$group": {"_id": {"customerNumber": "$customerNumber","customerName": "$customerName"}, #regroupement par numéro client et nom client
                "total_payé": {"$sum": {"$sum": "$orders.payments.amount"}},
                "total_commande": {"$sum": {"$multiply": ["$orders.order_details.quantityOrdered",
                                                        "$orders.order_details.priceEach"]}}}},

    {"$group": {"_id": "$_id.customerNumber", 
                "customerName": {"$first": "$_id.customerName"},
                "total_payé": {"$sum": "$total_payé"},
                "total_commande": {"$sum": "$total_commande"}}},

    {"$match": {"$expr": {"$gt": ["$total_payé", "$total_commande"]}}},

    {"$project": {"_id": 0,
                 "customerNumber": "$_id",
                 "customerName": 1,
                 "total_payé": 1,
                 "total_commande": 1}}
])))

customerName  total_payé  total_commande  customerNumber
0         Auto Associés & Cie.  1167017.40        64834.30             256
1           Baane Mini Imports  3731174.08       116599.19             121
2                 Royale Belge   267520.80        33440.10             381
3           Amica Models & Co.  2447048.76        94117.26             249
4   Men 'R' US Retailers, Ltd.   672678.44        48048.46             347
..                         ...         ...             ...             ...
93            FunGiftIdeas.com  2572018.02        98923.77             462
94        Technics Stores Inc.  4106624.38       120783.07             161
95   Norway Gifts By Mail, Co.  1901381.52        79224.23             299
96        Blauer See Auto, Co.  1873774.98        85171.59             128
97             Mini Auto Werke   783958.50        52263.90             452

[98 rows x 4 columns]